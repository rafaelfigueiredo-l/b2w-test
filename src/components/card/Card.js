import React, { Component } from "react";
import styled from "styled-components";
import Api from "../../service/SwapiApiService";
import CardButton from "./CardButton";
import Loader from "./CardLoader";

class Card extends Component {
  state = {
    planetId: null,
    cardData: null
  };

  componentDidMount() {
    this.getRandomPlanet();
  }

  render() {
    if (this.state.cardData === null) {
      return (
        <CardLayout id="Card">
          <CardBox>
            <Loader />
          </CardBox>
        </CardLayout>
      );
    }

    const {cardData} = this.state;
    const { name } = cardData;
    return (
      <CardLayout id="Card">
        <CardBox>
          <CardData>
            <CardName>
              <h2>{name.toUpperCase()}</h2>
            </CardName>
              <ListData>
                {this.printCardData(cardData)}
              </ListData>
          </CardData>
          <CardButton getRandomPlanet={this.getRandomPlanet} />
        </CardBox>
      </CardLayout>
    );
  }

  getRandomPlanet = async id => {
    this.setState({ planetId: null, cardData: null });
    const data = await Api.getRandomPlanet();
    const { name, population, climate, terrain, films } = data;
    this.setState({
      planetId: id,
      cardData: {
        name,
        population,
        climate,
        terrain,
        films
      }
    });
  };

  printCardData = (data) =>{

    const arrEntries = Object.entries(data);

    const list = arrEntries.map((el,i) => {
      const key = arrEntries[i][0].toString().toUpperCase();
      const listItemValue = arrEntries[i][1].toString().toUpperCase();


      if('NAME' === key){
        return false;
      }

      if('FILMS' === key){
        const filmsCount = arrEntries[i][1].length;

        const listFilmValue =
        (filmsCount > 1) ? `Featured In: ${filmsCount} Films` :
        (filmsCount === 0 ) ? 'Not Featured in Films':
        (filmsCount === 1) ? `Featured In: ${filmsCount} Film`:
        '';
        return (
          <li className="list-item film-item" key={i}>
            <span>{listFilmValue.toUpperCase()}</span>
          </li>
        );
      }

      return (
        <li className="list-item" key={i}>
          <span>{key}:</span> {listItemValue}
        </li>
      );
    });

    return list;

  }
}

const CardLayout = styled.section`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  `;

  const CardBox = styled.div`
  width:300px;
`;

const CardName = styled.div`
  padding: 10px 20px;
  border-bottom:2px solid #000;
`;
const CardData = styled.div`
  background: #fff;
  border: 10px solid #FFE919;
  width:100%;
`;

const ListData = styled.ul`
  padding: 10px 20px;
  list-style: none;
  span {
    font-weight: bold;
  }
  .film-item {
    text-align: center;
    font-size: 0.9em;
    text-align: center;
    padding-top: 30px;
  }
`;
export default Card;

