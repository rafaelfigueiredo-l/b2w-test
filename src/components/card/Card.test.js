import Card from './Card';

const func = () => {
    return true;
};

describe('<Card />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Card setAlert={func} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Snapshot testing', () => {
    const wrapper = shallow(<Card setAlert={func} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null if state.error is true', () => {
    const wrapper = shallow(<Card setAlert={func} />);
    wrapper.setState({ error: true });
    expect(wrapper).toEqual({});
  });

  it('should render Loader if state.cardData is null', () => {
    const wrapper = shallow(<Card setAlert={func} />);
    wrapper.setState({cardData: null});
    expect(wrapper.find('Loader')).toHaveLength(1);
  });

  it('should render h2 if state.cardData is not null', () => {
    const wrapper = shallow(<Card setAlert={func} />);
    wrapper.setState({cardData: {
      name: "Kalee",
      population: "4000000000",
      climate: "arid, temperate, tropical",
      terrain: "rainforests, cliffs, canyons, seas",
      films: [],
    }});
    expect(wrapper.find('h2')).toHaveLength(1);
  });

});