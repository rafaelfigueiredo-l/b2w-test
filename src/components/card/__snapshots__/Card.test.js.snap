// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Snapshot testing 1`] = `
<styled.section
  id="Card"
>
  <styled.div>
    <Loader />
  </styled.div>
</styled.section>
`;
